### PR TITLE
[Team 2] Update agent

### PR DESCRIPTION
**Pull request by tomeaton17**
# Summary

Final agent using WoLF-PHC with communication.
## Additional Information

Custom logging also implemented, thanks to @JaafarRammal :), but would not recommend using as log is 10MB+.
## Test Plan
Written unit tests for our agent and testing simulation with other agents.

### CODE REVIEW

File: pkg/agents/team2/message.go [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+func (a *CustomAgent2) AskNeighbourHP() {
+	msg := messages.NewAskHPMessage(a.ID(), a.Floor(), a.Floor()+1)
+	a.SendMessage(msg)
+}
+
+func (a *CustomAgent2) CheckNeighbourHP() {
+	msg := a.ReceiveMessage()
+	for msg != nil {
+		if msg.MessageType() == 7 {
```
Comment (jzzheng22) [inline pkg/agents/team2/message.go:13]: This is not safe - you should use `messages.MESSAGE_TYPE_HERE` instead. The numbers may change (e..g one of the types was removed overnight)

File: pkg/agents/team2/message.go [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+		}
+		msg = a.ReceiveMessage()
+	}
+}
+
+func (a *CustomAgent2) ReplyAllAskMsg() {
+	msg := a.ReceiveMessage()
+	for msg != nil {
+		msgType := msg.MessageType()
+		if msgType == 1 || msgType == 2 || msgType == 4 {
```
Comment (jzzheng22) [inline pkg/agents/team2/message.go:24]: See comment above - use MessageType comparison

File: pkg/agents/team2/policies.go [hunk 3]
```diff
@@ -41,3 +55,47 @@ func (a *CustomAgent2) adjustPolicies() {
...
+	if sumPolicies > sumAveragePolicies {
+		return Delta_win
+	}
+	return Delta_lose
+}
+
+func (a *CustomAgent2) exportPolicies() {
+	f, err := os.OpenFile(fmt.Sprintf("%s%s%s", a.ID(), "policies", ".csv"), os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0600)
+	if err != nil {
+		panic(err)
```
Comment (jzzheng22) [inline pkg/agents/team2/policies.go:77]: Do not use panic. Throw an error if you're going to do this, but don't use panic.

File: pkg/agents/team2/policies.go [hunk 4]
```diff
@@ -41,3 +55,47 @@ func (a *CustomAgent2) adjustPolicies() {
...
+	for i := 0; i < len(policies); i++ {
+		for j := 0; j < len(policies[0]); j++ {
+			sPolicies[i][j] = fmt.Sprint(policies[i][j])
+		}
+	}
+
+	err = w.WriteAll(sPolicies)
+
+	if err != nil {
+		log.Fatalln("error writing csv:", err)
```
Comment (jzzheng22) [inline pkg/agents/team2/policies.go:99]: Does it have to be fatal?

File: pkg/agents/team2/qtable.go [hunk 5]
```diff
@@ -1,7 +1,11 @@
 package team2
 
 import (
-	"math"
+	"encoding/csv"
+	"fmt"
+	"log"
+	"os"
+	//"time"
```
Comment (jzzheng22) [inline pkg/agents/team2/qtable.go:8]:
```suggestion
```

File: pkg/agents/team2/qtable.go [hunk 6]
```diff
@@ -36,3 +40,33 @@ func (a *CustomAgent2) updateQTable(state int, action int) {
 	a.qTable[state][action] = (1-Alpha)*a.qTable[state][action] +
 		Alpha*(a.rTable[state][action]+Gamma*a.getMaxQ(a.CheckState()).maxQ)
 }
+
+func (a *CustomAgent2) exportQTable() {
+	f, err := os.OpenFile(fmt.Sprintf("%s%s%s", a.ID(), "qtable", ".csv"), os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0600)
+	if err != nil {
+		panic(err)
```
Comment (jzzheng22) [inline pkg/agents/team2/qtable.go:47]: See above comment: don't use panic

File: pkg/agents/team2/qtable.go [hunk 7]
```diff
@@ -36,3 +40,33 @@ func (a *CustomAgent2) updateQTable(state int, action int) {
...
+		for j := 0; j < len(qTable[0]); j++ {
+			sQTable[i][j] = fmt.Sprint(qTable[i][j])
+		}
+	}
+
+	//fmt.Println("rows:", len(sQTable), "columns:", len(sQTable[0]))
+	err = w.WriteAll(sQTable)
+
+	if err != nil {
+		log.Fatalln("error writing csv:", err)
```
Comment (jzzheng22) [inline pkg/agents/team2/qtable.go:70]: See above comment - does it have to be fatal?

**Review by jzzheng22 — CHANGES_REQUESTED** [review]
Thanks, couple of changes that need to be addressed

File: pkg/agents/team2/message.go (hunk 1 above)
Comment (jzzheng22) [reply pkg/agents/team2/message.go:13]: Example: https://github.com/SOMAS2021/SOMAS2021/blob/main/pkg/agents/team4/finalAgent/messages.go#L15

File: pkg/agents/team2/config.go [hunk 8]
```diff
@@ -77,58 +73,111 @@ func InitTable(numStates int, numActions int) [][]float64 {
...
+		a.updatePolicies(oldState, action)
+		a.lastAge = a.Age()
+	}
+
+}
+
+func (a *CustomAgent2) isNewDay() bool {
+	if a.Age() < a.lastAge {
+		a.lastAge = -1
+		return true
```
[the remark this answers is not in the document]
Comment (tomeaton17) [reply pkg/agents/team2/config.go:148]: I believe this was added for some cases introduced by our reincarnation modification.

File: pkg/agents/team2/actions_test.go [hunk 9]
```diff
@@ -0,0 +1,35 @@
+package team2
+
+import (
+	"testing"
+)
+
+/* TODO: Implement this test. This is old implementation, do not use
```
Comment (SaraFFdez) [inline pkg/agents/team2/actions_test.go:7]: Could this commented out code be deleted?

File: pkg/agents/team2/message.go (hunk 1 above)
Comment (tomeaton17) [reply pkg/agents/team2/message.go:13]: Will fix shortly.

File: pkg/agents/team2/config.go [hunk 10]
```diff
@@ -77,58 +73,111 @@ func InitTable(numStates int, numActions int) [][]float64 {
...
-	stateSpace := InitStateSpace(3, 3, 3)
-	actionSpace := InitActionSpace()
-	policies := InitPolicies(27, 3)
-	rTable := InitTable(27, 3)
-	qTable := InitTable(27, 3)
+	actionDim := 6
+	daysAtCriticalDim := baseAgent.HealthInfo().MaxDayCritical + 1
+	stateSpace := InitStateSpace(10, 10, daysAtCriticalDim, 11)
+	actionSpace := InitActionSpace(actionDim)
+	policies := InitPolicies(10*10*daysAtCriticalDim*11, actionDim)
```
Comment (SaraFFdez) [inline pkg/agents/team2/config.go:80]:
```suggestion
	policies := InitPolicies(1100*daysAtCriticalDim, actionDim)
```
A suggestion. Similar in other variables

File: pkg/agents/team2/message.go (hunk 2 above)
Comment (tomeaton17) [reply pkg/agents/team2/message.go:24]: Will fix shortly.

File: pkg/agents/team2/policies.go (hunk 4 above)
Comment (tomeaton17) [reply pkg/agents/team2/policies.go]: Fixed

File: pkg/agents/team2/policies.go (hunk 3 above)
Comment (tomeaton17) [reply pkg/agents/team2/policies.go]: Fixed

File: pkg/agents/team2/qtable.go (hunk 5 above)
Comment (tomeaton17) [reply pkg/agents/team2/qtable.go]: Fixed

File: pkg/agents/team2/qtable.go (hunk 6 above)
Comment (tomeaton17) [reply pkg/agents/team2/qtable.go]: Fixed

File: pkg/agents/team2/qtable.go (hunk 7 above)
Comment (tomeaton17) [reply pkg/agents/team2/qtable.go]: Fixed

File: pkg/agents/team2/config.go [hunk 11]
```diff
@@ -77,58 +73,111 @@ func InitTable(numStates int, numActions int) [][]float64 {
...
-	//Perform the following only when platform arrives
-	//NOTE: should let infra team add a func to see whether the plaftfrom has arrived or not
-	if a.CheckState() != -1 {
-		a.isPlatArrived = true
-	} else {
-		a.isPlatArrived = false
-	}
-	if a.isPlatArrived {
+	//Perform the following only once per day when platform arrives
+	//a.exportQTable()
```
Comment (SaraFFdez) [inline pkg/agents/team2/config.go:111]: note this is commented out. Might be a good idea to delete comments that you will not use.

File: pkg/agents/team2/policies.go [hunk 12]
```diff
@@ -41,3 +55,47 @@ func (a *CustomAgent2) adjustPolicies() {
...
+	if sumPolicies > sumAveragePolicies {
+		return Delta_win
+	}
+	return Delta_lose
+}
+
+func (a *CustomAgent2) exportPolicies() {
+	f, err := os.OpenFile(fmt.Sprintf("%s%s%s", a.ID(), "policies", ".csv"), os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0600)
+	if err != nil {
+		log.Println("error opening csv: ", err)
```
Comment (jzzheng22) [inline pkg/agents/team2/policies.go:77]: Ideally use `log.Error` I think. Will defer to @JaafarRammal  over this though

File: pkg/agents/team2/config.go (hunk 11 above)
Comment (SaraFFdez) [inline pkg/agents/team2/config.go:111]: Note these functions are commented out. Might be a good idea to delete if they wont be used.

File: pkg/agents/team2/qtable.go [hunk 13]
```diff
@@ -36,3 +39,33 @@ func (a *CustomAgent2) updateQTable(state int, action int) {
 	a.qTable[state][action] = (1-Alpha)*a.qTable[state][action] +
 		Alpha*(a.rTable[state][action]+Gamma*a.getMaxQ(a.CheckState()).maxQ)
 }
+
+func (a *CustomAgent2) exportQTable() {
+	f, err := os.OpenFile(fmt.Sprintf("%s%s%s", a.ID(), "qtable", ".csv"), os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0600)
+	if err != nil {
+		log.Println("error opening csv: ", err)
```
Comment (jzzheng22) [inline pkg/agents/team2/qtable.go:46]: Ideally use `log.Error` I think. Will defer to @JaafarRammal  over this though

File: pkg/agents/team2/qtable.go [hunk 14]
```diff
@@ -36,3 +39,33 @@ func (a *CustomAgent2) updateQTable(state int, action int) {
...
+		for j := 0; j < len(qTable[0]); j++ {
+			sQTable[i][j] = fmt.Sprint(qTable[i][j])
+		}
+	}
+
+	//fmt.Println("rows:", len(sQTable), "columns:", len(sQTable[0]))
+	err = w.WriteAll(sQTable)
+
+	if err != nil {
+		log.Println("error writing csv:", err)
```
Comment (jzzheng22) [inline pkg/agents/team2/qtable.go:69]: Ideally use `log.Error` I think. Will defer to @JaafarRammal  over this though

File: pkg/agents/team2/policies.go [hunk 15]
```diff
@@ -41,3 +55,47 @@ func (a *CustomAgent2) adjustPolicies() {
...
+	for i := 0; i < len(policies); i++ {
+		for j := 0; j < len(policies[0]); j++ {
+			sPolicies[i][j] = fmt.Sprint(policies[i][j])
+		}
+	}
+
+	err = w.WriteAll(sPolicies)
+
+	if err != nil {
+		log.Println("error writing csv:", err)
```
Comment (jzzheng22) [inline pkg/agents/team2/policies.go:99]: Ideally use `log.Error` I think. Will defer to @JaafarRammal  over this though

File: pkg/agents/team2/config.go [hunk 16]
```diff
@@ -77,58 +73,111 @@ func InitTable(numStates int, numActions int) [][]float64 {
...
-			}
+			//if there's error, cease updating tables
+			return
 		}
 		a.Log("Agent team2:", infra.Fields{"selected and performed action": action})
 		a.Log("Agent team2 after action:", infra.Fields{"floor": a.Floor(), "hp": a.HP(), "food": a.CurrPlatFood(), "state": a.CheckState()})
+		if a.DaysAtCritical() > 0 && a.PreviousdayAtCritical != a.DaysAtCritical() {
+			a.PreviousdayAtCritical = a.DaysAtCritical()
+			a.daysAtCriticalCounter += 1
+			if a.DaysAtCritical() >= (a.HealthInfo().MaxDayCritical - 1) {
```
Comment (SaraFFdez) [inline pkg/agents/team2/config.go:131]: Will a.DaysAtCritical ever be bigger than MaxDayCritical - 1? (or is this just an extra check?)

**Review by woonmoon — COMMENTED** [review]
LGTM overall, nothing stylistically to say apart from what @jzzheng22 has already said. Just to confirm- you guys aren't sending any messages or treaties and any treaties you receive you reject on sight, is this correct?

File: pkg/agents/team2/config.go [hunk 17]
```diff
@@ -77,58 +73,111 @@ func InitTable(numStates int, numActions int) [][]float64 {
...
+func (a *CustomAgent2) isNewDay() bool {
+	if a.Age() < a.lastAge {
+		a.lastAge = -1
+		return true
+	}
+	if a.Age() == a.lastAge {
+		return false
 	}
+	return true
+}
```
Comment (SaraFFdez) [inline pkg/agents/team2/config.go:154]:
```suggestion
	return !(a.Age == a.lastAge)
}
```

File: pkg/agents/team2/policies.go (hunk 12 above)
Comment (tomeaton17) [reply pkg/agents/team2/policies.go:77]: `log.Error` is not a function.

File: pkg/agents/team2/config.go (hunk 17 above)
Comment (SaraFFdez) [reply pkg/agents/team2/config.go:154]: If you wanna reduce line count

**Comment by tomeaton17** [comment]
> LGTM overall, nothing stylistically to say apart from what @jzzheng22 has already said. Just to confirm- you guys aren't sending any messages or treaties and any treaties you receive you reject on sight, is this correct?

We send askHP and reply to all messages of type Ask but requests are ignored. Treaties are not implemented.

File: pkg/agents/team2/config.go [hunk 18]
```diff
@@ -77,58 +73,111 @@ func InitTable(numStates int, numActions int) [][]float64 {
...
+
+func (a *CustomAgent2) HandleAskIntendedFoodTaken(msg messages.AskIntendedFoodIntakeMessage) {
+	reply := msg.Reply(a.ID(), a.Floor(), msg.SenderFloor(), (a.lastAction+1)*5)
+	a.SendMessage(reply)
+	a.Log("Team2 replying askIntendedFoodTaken message from ", infra.Fields{"senderFloor": msg.SenderFloor(), "myFloor": a.Floor()})
+}
+
+func (a *CustomAgent2) HandleStateHP(msg messages.StateHPMessage) {
+	statement := msg.Statement()
+	a.neiboughHP = statement
```
Comment (SaraFFdez) [inline pkg/agents/team2/config.go:176]: This neighbour HP variable will be updated regardless of who sends the message even if it s from multiple floors above, is that intended?

Comment (SaraFFdez) [inline pkg/agents/team2/config.go:176]:
```suggestion
	a.neiboughHP =  msg.Statement()
```

**Review by SaraFFdez — CHANGES_REQUESTED** [review]
Looks good just some small changes to consider, thx!

File: pkg/agents/team2/config.go (hunk 16 above)
Comment (tomeaton17) [reply pkg/agents/team2/config.go:129]: Yes, MaxDayCritical is the threshold for the agent to enter a critical state. (The variable name could be clearer I agree)

File: pkg/agents/team2/config.go (hunk 18 above)
Comment (laiwenq) [reply pkg/agents/team2/config.go]: Thank you for pointing it out! Fixed it.

File: pkg/agents/team2/message.go (hunk 1 above)
Comment (laiwenq) [reply pkg/agents/team2/message.go]: Fixed.

File: pkg/agents/team2/message.go (hunk 2 above)
Comment (laiwenq) [reply pkg/agents/team2/message.go]: Fixed.

**Comment by tomeaton17** [comment]
@SaraFFdez @jzzheng22 everything you requested is fixed now so please can you approve :)

**Review by jzzheng22 — APPROVED** [review]
Thanks Tom - looks good. Ideally would prefer if you could wait for Jaafar to confirm whether the error handling after opening the file is acceptable

**Comment by tomeaton17** [comment]
> Thanks Tom - looks good. Ideally would prefer if you could wait for Jaafar to confirm whether the error handling after opening the file is acceptable

:) In that case whoever is the last person to accept the PR please merge it as well.

**Review by SaraFFdez — DISMISSED** [review]

**Comment by jzzheng22** [comment]
@tomeaton17  sorry - actualyl can you change something? The reason your `log.Error` didn't work si because you're not importing the right logging library.

Instead of importing `log`, can you import `log "github.com/sirupsen/logrus"`? Then change the errors from `log.Println` to `log.Error`?

File: pkg/agents/team2/policies.go (hunk 12 above)
Comment (JaafarRammal) [reply pkg/agents/team2/policies.go]: Yes please, `log.Error` (log is short name for logrus when we import it)

File: pkg/agents/team2/policies.go (hunk 15 above)
Comment (JaafarRammal) [reply pkg/agents/team2/policies.go]: yep

**Comment by tomeaton17** [comment]
> @tomeaton17 sorry - actualyl can you change something? The reason your `log.Error` didn't work si because you're not importing the right logging library.
> 
> Instead of importing `log`, can you import `log "github.com/sirupsen/logrus"`? Then change the errors from `log.Println` to `log.Error`?

Just committed this now.

**Review by JaafarRammal — COMMENTED** [review]
~~I suggest to add `log "github.com/sirupsen/logrus"` to your imports and to use `log.Error` in the mentioned comments. :) thanks!~~

Didn't realize you were doing it at the same time!

**Review by jzzheng22 — APPROVED** [review]
Thanks a lot!

**Comment by jzzheng22** [comment]
@tomeaton17  feel free to merge :)

**Pull request merged**